core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+            this.function = cast.function.and(function);
+        } else {
+            this.source = (StreamMessage<Object>) source;
+            this.function = (MapperFunction<Object, U>) function;
+        }
+    }
+
+    @VisibleForTesting
+    StreamMessage<Object> upstream() {
+        return source;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return source.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return source.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return source.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return source.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor) {
+        subscribe(subscriber, executor, EMPTY_OPTIONS);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new FuseableSubscriber<U>(subscriber, function), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class FuseableSubscriber<U> implements Subscriber<Object>, Subscription {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FuseableSubscriber, Subscription> upstreamUpdater =
+                AtomicReferenceFieldUpdater
+                        .newUpdater(FuseableSubscriber.class, Subscription.class, "upstream");
+
+        private final Subscriber<? super U> downstream;
+        private final MapperFunction<Object, U> function;
+
+        @Nullable
+        private volatile Subscription upstream;
+
+        FuseableSubscriber(Subscriber<? super U> downstream, MapperFunction<Object, U> function) {
+            requireNonNull(downstream, "downstream");
+            this.downstream = downstream;
+            this.function = function;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
+                subscription.cancel();
+                throw new IllegalStateException("Subscription already set!");
+            }
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(Object item) {
+            requireNonNull(item, "item");
+
+            final Subscription upstream = this.upstream;
+            if (upstream == null) {
```

Comment:
ditto.

Reply:
It means is there any chance that upstream is null?

Reply:
I mean that we usually do not have this logic in other subscribers. So if we need this, we should put this logic to other subscribers as well. 😄

Reply:
Other `Subscriber`s have `canceled` flags only to know the `Subscription` was canceled. However, I make ``upstream` null instead of adding an additional flag.
If this looks confusing, I can add a `canceled` flag for consistency.

Reply:
I missed the null setting. 😄 
Yeah, I think it's better to add the flag. We can fix the `request()` method as well.